pynih/source/python/type.d

```diff
@@ -913,25 +915,56 @@ private template PythonOpCmp(T) {
 
 
 private template PythonSubscript(T) {
+
     static extern(C) PyObject* _py_index(PyObject* self, PyObject* key) nothrow {
-        import python.raw: pyIndexCheck, pySliceCheck;
+        import python.raw: pyIndexCheck, pySliceCheck, PyObject_Repr, PyObject_Length,
+            Py_ssize_t, PySlice_GetIndices;
         import python.conv.python_to_d: to;
         import python.conv.d_to_python: toPython;
         import std.traits: Parameters, Unqual, hasMember;
+        import std.meta: Filter;
 
         PyObject* impl() {
             static if(hasMember!(T, "opIndex")) {
-                if(pyIndexCheck(self)) {
+                if(pyIndexCheck(key)) {
                     static if(__traits(compiles, Parameters!(T.opIndex))) {
                         alias parameters = Parameters!(T.opIndex);
                         static if(parameters.length == 1)
                             return self.to!(Unqual!T).opIndex(key.to!(parameters[0])).toPython;
                         else
                             throw new Exception("Don't know how to handle opIndex with more than one parameter");
-                    } else
-                        throw new Exception("Cannot determine parameters of " ~ T.stringof, ".opIndex");
+                    }
+                } else if(pySliceCheck(key)) {
+
+                    enum hasTwoParams(alias F) = Parameters!F.length == 2;
+                    alias twoParamOpSlices = Filter!(hasTwoParams, __traits(getOverloads, T, "opSlice"));
+
+                    static if(twoParamOpSlices.length > 0) {
+
+                        static assert(twoParamOpSlices.length == 1);
+                        alias opSlice = twoParamOpSlices[0];
+
+                        const len = PyObject_Length(self);
+                        Py_ssize_t start, stop, step;
+                        const indicesRet = PySlice_GetIndices(key, len, &start, &stop, &step);
+
+                        if(indicesRet < 0)
+                            throw new Exception("Could not get slice indices for key '" ~ PyObject_Repr(key).to!string ~ "'");
+
+                        if(step != 1)
+                            throw new Exception("Slice steps other than 1 not supported in D: " ~ PyObject_Repr(key).to!string);
+
+                        auto dObj = self.to!T;
+                        return dObj.opSlice(start, stop).toPython;
+
+                    } else {
+                        throw new Exception(T.stringof ~ " cannot be sliced by " ~ PyObject_Repr(key).to!string);
+                    }
+
+                    assert(0, "Error in slicing " ~ T.stringof ~ " with " ~ PyObject_Repr(key).to!string);
```

Comment:
Why not just
```D
                } else if (pySliceCheck(key)) {
                    static if (__traits(compiles, T.init[Py_ssize_t.init .. Py_ssize_t.init])) {
                        const len = PyObject_Length(self);
                        Py_ssize_t start, stop, step;
                        const indicesRet = PySlice_GetIndices(key, len, &start, &stop, &step);

                        if (indicesRet < 0)
                            throw new Exception("Could not get slice indices for key '" ~ PyObject_Repr(key).to!string ~ "'");

                        if (step != 1)
                            throw new Exception("Slice steps other than 1 not supported in D: " ~ PyObject_Repr(key).to!string);

                        return (self.to!T)[start .. stop].toPython;
                    } else
                        throw new Exception(T.stringof ~ " cannot be sliced by " ~ PyObject_Repr(key).to!string);
                    assert(0);
```

then you would support all possible types of 1 dimensional D slicing.

Reply:
I'm not sure I understand the suggestion.

Reply:
Calling opSlice doesn't necessarily give you a slice, what with there being two different ways of doing slice overloading in D.

So calling opSlice directly is the wrong approach, you should use the slice operators

Reply:
Ah, I see. Thanks!